shared/src/main/scala/mlscript/JSBackend.scala

```diff
@@ -70,6 +73,9 @@ class JSBackend {
       throw new Error(s"term $t is not a valid pattern")
   }
 
+  private var hasWithConstruct = false
+  private var withConstructFnName = "withConstruct"
```

Comment:
Why a `var`? This should be initialized to the right value directly. Initializing it with a wrong value is begging for errors to happen.

Reply:
They will be modified during the code generation, so I’m using `var`. We don’t know if we need the `withConstruct` helper function ahead of time.

Reply:
In this case you probably want to initialize it with `None` and later make it `Some(usedName)`. Anything you can to avoid having a variable that temporarily holds **_a wrong information_**.

Reply:
I can make `withConstructFnName` a `val`. But for `hasWithConstruct`, it will be modified during the traversal, so it must be `var` for now.

There are two ways making `hasWithConstruct` a `val`. One is add a recursive function which checks if `With` presents in the tree. Another one is to do the check during translation and append the value to return value. Both of them introduce more changes to this PR.

I’m really willing to improve the code quality concerning code generation. But I prefer doing it after I add code generation tests (#34) so that refactoring will not accidentally bring much regression.

Reply:
It's a bit better yes. But what I was suggesting was simpler and closer to the original intent (which was to avoid computing a name if it won't be needed).

To be clear what I was suggesting was this:
```diff
-  private var hasWithConstruct = false
-  private var withConstructFnName = "withConstruct"
+  private var withConstructFnName: Opt[Str] = N
```
and later, upon use, if `withConstructFnName` is empty just initialize it with the right value.
This way you restrict the state space to only valid situations.

Yet another, perhaps even better way would be:
```diff
  private var hasWithConstruct = false
-  private var withConstructFnName = "withConstruct"
+  private lazy val withConstructFnName = { hasWithConstruct = true; getAlternativeName("withConstruct") }
```